File: 2021/Day2/Day2.fs
module Day2

open System.IO

type Direction =
    | Forward of int
    | Down of int
    | Up of int

let parse (text: string) =
    match text.Split [| ' ' |] with
    | [| "forward"; v |] -> Forward(v |> int)
    | [| "down"; v |] -> Down(v |> int)
    | [| "up"; v |] -> Up(v |> int)
    | _ -> failwith $"Bad input: {text}"

let dive (hor, dep) curr =
    match curr with
    | Forward f -> (hor + f, dep)
    | Down d -> (hor, dep + d)
    | Up u -> (hor, dep - u)

let firstPart input = input |> Array.fold dive (0, 0)

let diveAim (hor, dep, aim) curr =
    match curr with
    | Forward f -> (hor + f, dep + f * aim, aim)
    | Down d -> (hor, dep, aim + d)
    | Up u -> (hor, dep, aim - u)

let secondPart input = input |> Array.fold diveAim (0, 0, 0)

let input =
    File.ReadAllLines "input.txt" |> Array.map parse

let (h1, d1) = firstPart input
let (h2, d2, _) = secondPart input

printfn $"First Part {h1 * d1}"
printfn $"Second Part {h2 * d2}"

File: 2021/Day02/Day2.scala
import scala.io.Source

@main def day2 =
    enum Direction:
        case Forward(value: Int)
        case Down(value: Int)
        case Up(value: Int)

    import Direction._

    def parse(text: String) =
        text.split(" ").toList match
            case "forward" :: v :: Nil => Forward(v.toInt)
            case "down" :: v :: Nil => Down(v.toInt)
            case "up" :: v :: Nil => Up(v.toInt)
            case _ => sys.error(s"Bad input: ${text}")

    def dive(acc: (Int, Int), curr: Direction) =
        val (hor, dep) = acc
        curr match
            case Forward(f) => (hor + f, dep)
            case Down(d)    => (hor, dep + d)
            case Up(u)      => (hor, dep - u)

    def diveAim(acc: (Int, Int, Int), curr: Direction) =
        val (hor, dep, aim) = acc
        curr match
            case Forward(f) => (hor + f, dep + f * aim, aim)
            case Down(d)    => (hor, dep, aim + d)
            case Up(u)      => (hor, dep, aim - u)

    val input = Source
        .fromFile("input.txt")
        .getLines
        .map(parse(_))
        .toList

    val (h1, d1) = input
        .foldLeft((0, 0))(dive)

    val (h2, d2, _) = input
        .foldLeft((0, 0, 0))(diveAim)

    println(s"Part 1: ${h1 * d1}")
    println(s"Part 2: ${h2 * d2}")
